File: src/lib/lunar.ts
// Moonbug lunar calculations (UTC-based), kept in sync with the Go backend.
export const SYNODIC_MONTH = 29.530588853;
const REF_NEW_MOON = 947166000; // Date.UTC(2000, 0, 6, 18, 14, 0) in seconds

export interface MoonStatus {
  age: number;
  illumination: number;
  phase: string;
  phaseCode: string;
  phaseEmoji: string;
  daysUntilFull: number;
  daysUntilNew: number;
}

const PHASES = [
  { name: "New Moon", emoji: "🌑", code: "new-moon" },
  { name: "Waxing Crescent", emoji: "🌒", code: "waxing-crescent" },
  { name: "First Quarter", emoji: "🌓", code: "first-quarter" },
  { name: "Waxing Gibbous", emoji: "🌔", code: "waxing-gibbous" },
  { name: "Full Moon", emoji: "🌕", code: "full-moon" },
  { name: "Waning Gibbous", emoji: "🌖", code: "waning-gibbous" },
  { name: "Last Quarter", emoji: "🌗", code: "last-quarter" },
  { name: "Waning Crescent", emoji: "🌘", code: "waning-crescent" },
];

function ageFor(date: Date): number {
  const days = (date.getTime() / 1000 - REF_NEW_MOON) / 86400;
  let age = days % SYNODIC_MONTH;
  if (age < 0) age += SYNODIC_MONTH;
  return age;
}

export function getMoonStatus(date: Date = new Date()): MoonStatus {
  const age = ageFor(date);
  const pct = age / SYNODIC_MONTH;
  const angle = pct * 2 * Math.PI;
  const illumination = ((1 - Math.cos(angle)) / 2) * 100;

  let idx = 0;
  if (pct < 0.03 || pct >= 0.97) idx = 0;
  else if (pct < 0.22) idx = 1;
  else if (pct < 0.28) idx = 2;
  else if (pct < 0.47) idx = 3;
  else if (pct < 0.53) idx = 4;
  else if (pct < 0.72) idx = 5;
  else if (pct < 0.78) idx = 6;
  else idx = 7;

  const phaseForDate = (d: Date) => {
    const a = ageFor(d);
    const p = a / SYNODIC_MONTH;
    let i = 0;
    if (p < 0.03 || p >= 0.97) i = 0;
    else if (p < 0.22) i = 1;
    else if (p < 0.28) i = 2;
    else if (p < 0.47) i = 3;
    else if (p < 0.53) i = 4;
    else if (p < 0.72) i = 5;
    else if (p < 0.78) i = 6;
    else i = 7;
    return i;
  };

  const daysUntilNext = (target: number) => {
    let delta = target - pct;
    if (delta < 0) delta += 1;
    return delta * SYNODIC_MONTH;
  };

  return {
    age,
    illumination,
    phase: PHASES[idx].name,
    phaseCode: PHASES[idx].code,
    phaseEmoji: PHASES[idx].emoji,
    daysUntilFull: daysUntilNext(0.5),
    daysUntilNew: daysUntilNext(0),
  };
}

export function phaseInfoForDate(date: Date) {
  const a = ageFor(date);
  const p = a / SYNODIC_MONTH;
  const angle = (p) * 2 * Math.PI;
  const illumination = ((1 - Math.cos(angle)) / 2) * 100;
  let idx = 0;
  if (p < 0.03 || p >= 0.97) idx = 0;
  else if (p < 0.22) idx = 1;
  else if (p < 0.28) idx = 2;
  else if (p < 0.47) idx = 3;
  else if (p < 0.53) idx = 4;
  else if (p < 0.72) idx = 5;
  else if (p < 0.78) idx = 6;
  else idx = 7;
  return { ...PHASES[idx], illumination: Math.round(illumination) };
}

export { PHASES };

// Phase fraction 0..1 used to render the lit portion of the moon SVG.
export function phaseFraction(age: number): number {
  return (age % SYNODIC_MONTH) / SYNODIC_MONTH;
}

File: backend/internal/lunar/lunar.go
package lunar

import (
	"math"
	"time"
)

const (
	SynodicMonth = 29.530588853
	RefNewMoon   = 947166000 // Date.UTC(2000, 0, 6, 18, 14, 0) in seconds
)

// PhaseName returns the canonical moon phase for a given lunar age in days.
func PhaseName(age float64) string {
	pct := age / SynodicMonth
	switch {
	case pct < 0.03 || pct >= 0.97:
		return "New Moon"
	case pct < 0.22:
		return "Waxing Crescent"
	case pct < 0.28:
		return "First Quarter"
	case pct < 0.47:
		return "Waxing Gibbous"
	case pct < 0.53:
		return "Full Moon"
	case pct < 0.72:
		return "Waning Gibbous"
	case pct < 0.78:
		return "Last Quarter"
	default:
		return "Waning Crescent"
	}
}

func PhaseEmoji(age float64) string {
	pct := age / SynodicMonth
	switch {
	case pct < 0.03 || pct >= 0.97:
		return "🌑"
	case pct < 0.22:
		return "🌒"
	case pct < 0.28:
		return "🌓"
	case pct < 0.47:
		return "🌔"
	case pct < 0.53:
		return "🌕"
	case pct < 0.72:
		return "🌖"
	case pct < 0.78:
		return "🌗"
	default:
		return "🌘"
	}
}

func PhaseCode(age float64) string {
	pct := age / SynodicMonth
	switch {
	case pct < 0.03 || pct >= 0.97:
		return "new-moon"
	case pct < 0.22:
		return "waxing-crescent"
	case pct < 0.28:
		return "first-quarter"
	case pct < 0.47:
		return "waxing-gibbous"
	case pct < 0.53:
		return "full-moon"
	case pct < 0.72:
		return "waning-gibbous"
	case pct < 0.78:
		return "last-quarter"
	default:
		return "waning-crescent"
	}
}

// RefNewMoonTime returns the reference new moon as a time.Time.
func RefNewMoonTime() time.Time {
	return time.Unix(RefNewMoon, 0).UTC()
}
func Age(t time.Time) float64 {
	secs := float64(t.Unix())
	days := (secs - float64(RefNewMoon)) / 86400.0
	age := math.Mod(days, SynodicMonth)
	if age < 0 {
		age += SynodicMonth
	}
	return age
}

// Illumination returns the percentage of the moon illuminated (0-100).
func Illumination(age float64) float64 {
	angle := (age / SynodicMonth) * 2 * math.Pi
	return ((1 - math.Cos(angle)) / 2) * 100
}

// DaysUntilNext returns the number of days until the next target phase.
// target 0 = New Moon, 0.5 = Full Moon (fraction of synodic month).
func DaysUntilNext(age, target float64) float64 {
	pct := age / SynodicMonth
	delta := target - pct
	if delta < 0 {
		delta += 1
	}
	return delta * SynodicMonth
}
